File: ios-renderer/Sources/A2UIRenderer/Models/Surface.swift
import Foundation
import Observation

/// Reactive surface state. When MessageProcessor updates properties, SwiftUI auto-re-renders.
@Observable
public final class Surface {
    public let surfaceId: String
    public var rootComponentId: String = "root"
    public var components: [String: A2UIComponent] = [:]
    public var dataModel: [String: JSONValue] = [:]
    public var designTokens: [String: DesignToken] = [:]

    public init(surfaceId: String) {
        self.surfaceId = surfaceId
    }

    /// Get a component by ID.
    public func component(id: String) -> A2UIComponent? {
        components[id]
    }

    /// Get child component IDs for a given component.
    public func childIds(for component: A2UIComponent) -> [String] {
        component.children?.explicitList ?? []
    }
}

File: android-renderer/lib/src/main/kotlin/com/interfaceai/a2ui/models/Surface.kt
package com.interfaceai.a2ui.models

import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.serialization.json.JsonElement

/**
 * Reactive surface state for Compose.
 * Uses mutableStateMapOf/mutableStateOf so Compose auto-recomposes when state changes.
 */
class Surface(val surfaceId: String) {
    var rootComponentId by mutableStateOf("root")
    val components = mutableStateMapOf<String, A2UIComponent>()
    var dataModel by mutableStateOf<Map<String, JsonElement>>(emptyMap())
    var designTokens by mutableStateOf<Map<String, DesignToken>>(emptyMap())

    fun component(id: String): A2UIComponent? = components[id]

    fun childIds(component: A2UIComponent): List<String> {
        return component.children?.explicitList ?: emptyList()
    }
}
